pilot/pkg/xds/ads.go

```diff
@@ -524,46 +523,38 @@ func (s *DiscoveryServer) DeltaAggregatedResources(stream discovery.AggregatedDi
 // for large configs. The method will hold a lock on con.pushMutex.
 func (s *DiscoveryServer) pushConnection(con *Connection, pushEv *Event) error {
 	pushRequest := pushEv.pushRequest
-	currentVersion := versionInfo()
-	// TODO: update the service deps based on NetworkScope
-	if !pushRequest.Full {
-		if !ProxyNeedsPush(con.proxy, pushEv) {
-			adsLog.Debugf("Skipping EDS push to %v, no updates required", con.ConID)
-			return nil
-		}
-		if con.proxy.WatchedResources[v3.EndpointType] == nil {
+
+	if pushRequest.Full {
+		adsLog.Infof("Pushing %v", con.ConID)
+		// Update Proxy with current information.
+		if err := s.updateProxy(con.proxy, pushRequest.Push); err != nil {
 			return nil
 		}
-		// TODO allow partial updates to types other than EDS
-		return s.pushXds(con, pushRequest.Push, s.Generators[v3.EndpointType],
-			currentVersion, con.Watched(v3.EndpointType), pushRequest.ConfigsUpdated)
 	}
 
-	// Update Proxy with current information.
-	if err := s.updateProxy(con.proxy, pushRequest.Push); err != nil {
-		return nil
-	}
-
-	// This depends on SidecarScope updates, so it should be called after SetSidecarScope.
 	if !ProxyNeedsPush(con.proxy, pushEv) {
 		adsLog.Debugf("Skipping push to %v, no updates required", con.ConID)
-
-		reportAllEvents(s.StatusReporter, con.ConID, pushRequest.Push.Version, nil)
+		if pushRequest.Full {
+			// Only report for full versions, incremental pushes do not have a new version
+			reportAllEvents(s.StatusReporter, con.ConID, pushRequest.Push.Version, nil)
+		}
 		return nil
 	}
 
-	adsLog.Infof("Pushing %v", con.ConID)
+	currentVersion := versionInfo()
 
 	// Send pushes to all generators
 	// Each Generator is responsible for determining if the push event requires a push
 	for _, w := range getPushResources(con.proxy.WatchedResources) {
```

Comment:
The idea here is we go through each generator (~5) even on empty pushes. Those generators will filter out incremental pushes if they do not support them.

We can also explicitly have a map of "incremental" generators and add EDS and SDS there if preferred - but I think the perf impact of this is trivial

Reply:
I think it is OK to iterate.